### PR TITLE
Drop Flyweight pattern

### PR DESCRIPTION
**Pull request by greg0ire**
There is a growing need for types with parameters, and that pattern was
probably used out of concerns that are no longer relevant anyway.

[The documentation needs to be updated, but sits in the wrong repository :see_no_evil: ](https://www.doctrine-project.org/projects/doctrine-orm/en/2.11/cookbook/custom-mapping-types.html)

EDIT: actually there is also one in this repository: https://www.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/types.html#custom-mapping-types :thinking:

### CODE REVIEW

File: docs/en/reference/types.rst [hunk 1]
```diff
@@ -909,3 +909,55 @@ hook it into the database platform:
...
+
+To do that, you can obtain the ``TypeRegistry`` singleton from ``Type``
+and register your type in it::
+
+    <?php
+    Type::getTypeRegistry()->addType('encrypted', new EncryptedType(
+        'aes-256-cbc-hmac-sha256',
+        '/path/to/secret',
+        $secretStore
+    ));
```
Comment (greg0ire) [inline docs/en/reference/types.rst:945]: Although this paragraph is a how-to, the one above it is as well, and the DBAL docs do not have a cookbook/how-to section. Not sure what to do.

**Comment by derrabus** [comment]
Because this PR implements a new feature, it should target the 3.3.x branch.

**Comment by morozov** [comment]
Given that the types are no longer identified by their names, what's the point of the name?

File: docs/en/reference/types.rst [hunk 2]
```diff
@@ -909,3 +909,55 @@ hook it into the database platform:
...
+        {
+            return 'encrypted';
+        }
+    }
+
+To do that, you can obtain the ``TypeRegistry`` singleton from ``Type``
+and register your type in it::
+
+    <?php
+    Type::getTypeRegistry()->addType('encrypted', new EncryptedType(
```
Comment (morozov) [inline docs/en/reference/types.rst]: This is a bit confusing. Why does the user need to specify the `'encrypted'` name of the custom type twice: once in the type itself and the other time during the registration?

It looks like we should deprecate `Type::getName()` and make sure no code relies on it.

**Comment by greg0ire** [comment]
> Given that the types are no longer identified by their names, what's the point of the name?

If for some reason, you have 2 instances of `EncryptedType`, you will have to make `EncryptedType::getName()` dynamic because it's not possible to register 2 types with the same name, because a type _instance_ is still identified by its name.

File: docs/en/reference/types.rst (hunk 2 above)
Comment (greg0ire) [reply docs/en/reference/types.rst]: I don't know but it is indeed confusing, and apparently already challenged: https://github.com/doctrine/dbal/blob/34af8fbc885d4853f8a138cb8fe3dfa7b0626dcb/src/Types/Type.php#L107-L109

That comment was introduced in 2009 lol: 537c8e4951d1908fad375b4d613d4a74253d117d

Comment (greg0ire) [reply docs/en/reference/types.rst]: After looking into it, it seems used for reverse engineering: https://github.com/doctrine/dbal/blob/34af8fbc885d4853f8a138cb8fe3dfa7b0626dcb/src/Platforms/AbstractPlatform.php#L440-L482

https://github.com/doctrine/dbal/blob/34af8fbc885d4853f8a138cb8fe3dfa7b0626dcb/src/Schema/AbstractSchemaManager.php#L1298-L1314

**Comment by morozov** [comment]
I still have a feeling that it isn't the next step in refactoring the type system. If there's no clear understanding of whether the type names are needed, and the names are potentially related to the types being flyweights, I'd pause here and clean up the naming.

The need to define the type name twice and declare it in the class dynamically depending on the implementation is a footgun that IMO needs to be addressed first of all.

If you believe this PR is the right next step, we can proceed.

**Review by morozov — DISMISSED** [review]

**Comment by greg0ire** [comment]
@morozov there is no hurry for this PR, we can pause it. I noticed that there is no mention of `DC2Type` inside docs, this might be worth an explanation (in the sense of the website you showed me, https://documentation.divio.com). Once we have that, we can see if we can work around it, for instance by having the type registry "keys" be used instead of calling `getName()`. The current implementation of `TypeRegistry::lookupName` looks like [it would just work](https://3v4l.org/0OS9D). I didn't pay enough attention and didn't see it was implemented with `array_search` and an object, which I have never used like this in my life.

Consequently, there is no need to deprecate `lookupName` in this PR.

**Comment by morozov** [comment]
Now that the confusion about using the type name twice has been eliminated in https://github.com/doctrine/dbal/pull/5208, would it make sense to retarget this patch for `4.0.x`?

**Comment by greg0ire** [comment]
Yes it would, I was planning on doing it soon :tm:

File: docs/en/reference/types.rst [hunk 3]
```diff
@@ -905,3 +905,50 @@ hook it into the database platform:
...
+        public function __construct(
+            private string $cipher,
+            private string $secretName,
+            private SecretStore $secretStore
+        ) {
+        }
+
+        public function convertToPHPValue($value, AbstractPlatform $platform): string
+        {
+            return openssl_decrypt(
```
Comment (morozov) [inline docs/en/reference/types.rst]: IMO, encryption is a quite specific and sophisticated domain to be used to illustrate a concept of types. It's also not self-contain since it depends on a `SecretStore` which isn't declared in the example.

It's better to use a dummy example with a prefix that doesn't distract from the subject.

File: tests/Types/TypeRegistryTest.php [hunk 4]
```diff
@@ -97,9 +97,11 @@ public function testRegisterWithAlreadyRegisteredInstance(): void
         $newType = new TextType();
 
         $this->registry->register('some', $newType);
-
-        $this->expectException(Exception::class);
-        $this->registry->register('other', $newType);
+        $this->registry->register('sameButDifferent', $newType);
+        self::assertSame(
+            $this->registry->get('some'),
+            $this->registry->get('sameButDifferent')
```
Comment (morozov) [inline tests/Types/TypeRegistryTest.php]: "sameButDifferent" is really confusing to read. Why not just use some arbitrary keys like "type1" and "type2"? The fact that they are represented by the same object will be clear from the context and doesn't have to be expressed in the names.

File: docs/en/reference/types.rst (hunk 1 above)
Comment (morozov) [reply docs/en/reference/types.rst:945]: It's fine to keep it here.

**Review by morozov — COMMENTED** [review]
Code-wise looks good. Just a few nitpicks on documentation and tests.

File: docs/en/reference/types.rst (hunk 3 above)
Comment (greg0ire) [reply docs/en/reference/types.rst]: Ok, I tried to dumb it down to a less realistic but more simple use case.

**Review by morozov — APPROVED** [review]
